pageserver/src/tenant/timeline.rs

```diff
@@ -3245,7 +3245,31 @@ impl Timeline {
                         || contains_hole
                     {
                         // ... if so, flush previous layer and prepare to write new one
-                        new_layers.push(writer.take().unwrap().finish(prev_key.unwrap().next())?);
+
+                        let end_key = prev_key.unwrap().next();
+
+                        let w = writer.take().unwrap();
+
+                        // If an identical L1 layer already exists, no need to create a new one.
+                        //
+                        // This can happen if compaction is interrupted after it has already
+                        // created some or all of the L1 layers, but has not deleted the L0 layers
+                        // yet, so that on next compaction, we do the same work again.
+                        //
+                        // NOTE: this is racy, if there can be any other task that concurrently
+                        // creates L1 layers. Currently, there can be only one compaction task
+                        // running at any time, so this is fine.
+                        if self.layers.read().unwrap().contains(
```

Comment:
You are checking presence duplicates under read lock but it is immediately released after this check.
So in principle this layer can be removed by some other task (i.e. GC) right after this point. 
As a result the will be no such layer in layer map at all.

Can this really happen? Is resurrection of deleted layers ever possible? I am not sure. But at least there is a gap between checking for duplicate and insertion of new layer in layer map not protected by `layes` lock.